lib/shareProcessor.js

```diff
@@ -86,33 +86,56 @@ var ShareProcessor = module.exports = function ShareProcessor(config, logger){
         return fromGroup + ':' + toGroup + ':shares:' + blockHash;
     }
 
+    this.shareCacheKey = function(fromGroup, toGroup){
+        return fromGroup + ':' + toGroup + ':sharecache';
+    }
+
     var pendingBlocksKey = 'pendingBlocks';
     var foundBlocksKey = 'foundBlocks';
     var hashrateKey = 'hashrate';
     var balancesKey = 'balances';
 
-    this._handleShare = function(share){
-        var redisTx = _this.redisClient.multi();
-        var currentMs = Date.now();
+    this._handleShare = function(share, callback){
         var fromGroup = share.job.fromGroup;
         var toGroup = share.job.toGroup;
+        var blockHash = share.blockHash;
         var currentRound = _this.currentRoundKey(fromGroup, toGroup);
-        redisTx.hincrbyfloat(currentRound, share.workerAddress, share.difficulty);
+        var cacheKey = _this.shareCacheKey(fromGroup, toGroup);
 
-        var currentTs = Math.floor(currentMs / 1000);
-        redisTx.zadd(hashrateKey, currentTs, [fromGroup, toGroup, share.worker, share.difficulty, currentMs].join(':'));
+        _this.redisClient.sadd(cacheKey, blockHash, function(error, result){
+            if (error){
+                logger.error('Check share duplicated failed, error: ' + error);
+                callback(error);
+                return;
+            }
 
-        if (share.foundBlock){
-            var blockHash = share.blockHash;
-            var newKey = _this.roundKey(fromGroup, toGroup, blockHash);
-            var blockWithTs = blockHash + ':' + currentMs.toString();
+            if (result === 0){
+                logger.error('Ignore duplicated share');
+                callback('Duplicated share');
+                return;
+            }
 
-            redisTx.rename(currentRound, newKey);
-            redisTx.sadd(pendingBlocksKey, blockWithTs);
-            redisTx.hset(foundBlocksKey, blockHash, share.workerAddress)
-        }
-        redisTx.exec(function(error, _){
-            if (error) logger.error('Handle share failed, error: ' + error);
+            var redisTx = _this.redisClient.multi();
+            var currentMs = Date.now();
+            redisTx.hincrbyfloat(currentRound, share.workerAddress, share.difficulty);
+
+            var currentTs = Math.floor(currentMs / 1000);
+            redisTx.zadd(hashrateKey, currentTs, [fromGroup, toGroup, share.worker, share.difficulty, currentMs].join(':'));
+
+            if (share.foundBlock){
+                var blockHash = share.blockHash;
+                var newKey = _this.roundKey(fromGroup, toGroup, blockHash);
+                var blockWithTs = blockHash + ':' + currentMs.toString();
+
+                redisTx.rename(currentRound, newKey);
+                redisTx.sadd(pendingBlocksKey, blockWithTs);
+                redisTx.hset(foundBlocksKey, blockHash, share.workerAddress)
+                redisTx.del(cacheKey);
```

Comment:
the cacheKey is only removed when a corresponding block is found. Since it's not possible to set TTL on set, we should have a cleanup mechanism which empty this set, otherwise it might quickly become unnecessarily big.